In [1]:
import numpy as np
import pandas as pd

In [2]:
np.__version__

'1.19.5'

In [3]:
pd.__version__

'1.3.2'

In [4]:
car_price_df = pd.read_csv('./car_price_data.csv')

In [5]:
car_price_df

Make       Model  Year                Engine Fuel Type  Engine HP  \
0          BMW  1 Series M  2011     premium unleaded (required)      335.0   
1          BMW    1 Series  2011     premium unleaded (required)      300.0   
2          BMW    1 Series  2011     premium unleaded (required)      300.0   
3          BMW    1 Series  2011     premium unleaded (required)      230.0   
4          BMW    1 Series  2011     premium unleaded (required)      230.0   
...        ...         ...   ...                             ...        ...   
11909    Acura         ZDX  2012     premium unleaded (required)      300.0   
11910    Acura         ZDX  2012     premium unleaded (required)      300.0   
11911    Acura         ZDX  2012     premium unleaded (required)      300.0   
11912    Acura         ZDX  2013  premium unleaded (recommended)      300.0   
11913  Lincoln      Zephyr  2006                regular unleaded      221.0   

       Engine Cylinders Transmission Type      Driven_Wheels  Number of Doors  \
0                   6.0            MANUAL   rear wheel drive              2.0   
1                   6.0            MANUAL   rear wheel drive              2.0   
2                   6.0            MANUAL   rear wheel drive              2.0   
3                   6.0            MANUAL   rear wheel drive              2.0   
4                   6.0            MANUAL   rear wheel drive              2.0   
...                 ...               ...                ...              ...   
11909               6.0         AUTOMATIC    all wheel drive              4.0   
11910               6.0         AUTOMATIC    all wheel drive              4.0   
11911               6.0         AUTOMATIC    all wheel drive              4.0   
11912               6.0         AUTOMATIC    all wheel drive              4.0   
11913               6.0         AUTOMATIC  front wheel drive              4.0   

                             Market Category Vehicle Size  Vehicle Style  \
0      Factory Tuner,Luxury,High-Performance      Compact          Coupe   
1                         Luxury,Performance      Compact    Convertible   
2                    Luxury,High-Performance      Compact          Coupe   
3                         Luxury,Performance      Compact          Coupe   
4                                     Luxury      Compact    Convertible   
...                                      ...          ...            ...   
11909             Crossover,Hatchback,Luxury      Midsize  4dr Hatchback   
11910             Crossover,Hatchback,Luxury      Midsize  4dr Hatchback   
11911             Crossover,Hatchback,Luxury      Midsize  4dr Hatchback   
11912             Crossover,Hatchback,Luxury      Midsize  4dr Hatchback   
11913                                 Luxury      Midsize          Sedan   

       highway MPG  city mpg  Popularity   MSRP  
0               26        19        3916  46135  
1               28        19        3916  40650  
2               28        20        3916  36350  
3               28        18        3916  29450  
4               28        18        3916  34500  
...            ...       ...         ...    ...  
11909           23        16         204  46120  
11910           23        16         204  56670  
11911           23        16         204  50620  
11912           23        16         204  50920  
11913           26        17          61  28995  

[11914 rows x 16 columns]

In [6]:
car_price_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 11914 entries, 0 to 11913
Data columns (total 16 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Make               11914 non-null  object 
 1   Model              11914 non-null  object 
 2   Year               11914 non-null  int64  
 3   Engine Fuel Type   11911 non-null  object 
 4   Engine HP          11845 non-null  float64
 5   Engine Cylinders   11884 non-null  float64
 6   Transmission Type  11914 non-null  object 
 7   Driven_Wheels      11914 non-null  object 
 8   Number of Doors    11908 non-null  float64
 9   Market Category    8172 non-null   object 
 10  Vehicle Size       11914 non-null  object 
 11  Vehicle Style      11914 non-null  object 
 12  highway MPG        11914 non-null  int64  
 13  city mpg           11914 non-null  int64  
 14  Popularity         11914 non-null  int64  
 15  MSRP               11914 non-null  int64  
dtypes: float64(3), int64(5

In [7]:
bmw_df = car_price_df[car_price_df['Make'] == 'BMW']
print(bmw_df.MSRP.mean())

61546.76347305389


Therefore, the average price of BMW cars in the dataset is : **61546**.

In [8]:
car_price_df[car_price_df.Year >= 2015]

Make       Model  Year                Engine Fuel Type  Engine HP  \
32     FIAT  124 Spider  2017  premium unleaded (recommended)      160.0   
33     FIAT  124 Spider  2017  premium unleaded (recommended)      160.0   
34     FIAT  124 Spider  2017  premium unleaded (recommended)      160.0   
41      BMW    2 Series  2016     premium unleaded (required)      240.0   
42      BMW    2 Series  2016     premium unleaded (required)      240.0   
...     ...         ...   ...                             ...        ...   
11898   BMW          Z4  2015     premium unleaded (required)      300.0   
11899   BMW          Z4  2015     premium unleaded (required)      335.0   
11900   BMW          Z4  2016     premium unleaded (required)      300.0   
11901   BMW          Z4  2016     premium unleaded (required)      240.0   
11902   BMW          Z4  2016     premium unleaded (required)      335.0   

       Engine Cylinders Transmission Type     Driven_Wheels  Number of Doors  \
32                  4.0            MANUAL  rear wheel drive              2.0   
33                  4.0            MANUAL  rear wheel drive              2.0   
34                  4.0            MANUAL  rear wheel drive              2.0   
41                  4.0         AUTOMATIC  rear wheel drive              2.0   
42                  4.0         AUTOMATIC  rear wheel drive              2.0   
...                 ...               ...               ...              ...   
11898               6.0  AUTOMATED_MANUAL  rear wheel drive              2.0   
11899               6.0  AUTOMATED_MANUAL  rear wheel drive              2.0   
11900               6.0  AUTOMATED_MANUAL  rear wheel drive              2.0   
11901               4.0            MANUAL  rear wheel drive              2.0   
11902               6.0  AUTOMATED_MANUAL  rear wheel drive              2.0   

               Market Category Vehicle Size Vehicle Style  highway MPG  \
32                 Performance      Compact   Convertible           35   
33                 Performance      Compact   Convertible           35   
34                 Performance      Compact   Convertible           35   
41          Luxury,Performance      Compact         Coupe           35   
42                      Luxury      Compact   Convertible           34   
...                        ...          ...           ...          ...   
11898  Luxury,High-Performance      Compact   Convertible           24   
11899  Luxury,High-Performance      Compact   Convertible           24   
11900  Luxury,High-Performance      Compact   Convertible           24   
11901       Luxury,Performance      Compact   Convertible           34   
11902  Luxury,High-Performance      Compact   Convertible           24   

       city mpg  Popularity   MSRP  
32           26         819  27495  
33           26         819  24995  
34           26         819  28195  
41           23        3916  32850  
42           23        3916  38650  
...         ...         ...    ...  
11898        17        3916  56950  
11899        17        3916  65800  
11900        17        3916  57500  
11901        22        3916  49700  
11902        17        3916  66350  

[5995 rows x 16 columns]

In [9]:
car_price_df[car_price_df.Year >= 2015].info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 5995 entries, 32 to 11902
Data columns (total 16 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Make               5995 non-null   object 
 1   Model              5995 non-null   object 
 2   Year               5995 non-null   int64  
 3   Engine Fuel Type   5995 non-null   object 
 4   Engine HP          5944 non-null   float64
 5   Engine Cylinders   5987 non-null   float64
 6   Transmission Type  5995 non-null   object 
 7   Driven_Wheels      5995 non-null   object 
 8   Number of Doors    5990 non-null   float64
 9   Market Category    4671 non-null   object 
 10  Vehicle Size       5995 non-null   object 
 11  Vehicle Style      5995 non-null   object 
 12  highway MPG        5995 non-null   int64  
 13  city mpg           5995 non-null   int64  
 14  Popularity         5995 non-null   int64  
 15  MSRP               5995 non-null   int64  
dtypes: float64(3), int64(5

We have a total of 5995 entries with year >= 2015. And using the df.info() method we find that there are 5944 non-null values in the Engine HP column. Therefore number of missing values for Engine HP = 5995 - 5944

In [10]:
car_price_df[(car_price_df.Year >= 2015)]['Engine HP'].isnull().sum()

51

Therefore, number of missing values for Engine HP after the year 2015 = **51**.

In [11]:
car_price_df.describe()

Year    Engine HP  Engine Cylinders  Number of Doors  \
count  11914.000000  11845.00000      11884.000000     11908.000000   
mean    2010.384338    249.38607          5.628829         3.436093   
std        7.579740    109.19187          1.780559         0.881315   
min     1990.000000     55.00000          0.000000         2.000000   
25%     2007.000000    170.00000          4.000000         2.000000   
50%     2015.000000    227.00000          6.000000         4.000000   
75%     2016.000000    300.00000          6.000000         4.000000   
max     2017.000000   1001.00000         16.000000         4.000000   

        highway MPG      city mpg    Popularity          MSRP  
count  11914.000000  11914.000000  11914.000000  1.191400e+04  
mean      26.637485     19.733255   1554.911197  4.059474e+04  
std        8.863001      8.987798   1441.855347  6.010910e+04  
min       12.000000      7.000000      2.000000  2.000000e+03  
25%       22.000000     16.000000    549.000000  2.100000e+04  
50%       26.000000     18.000000   1385.000000  2.999500e+04  
75%       30.000000     22.000000   2009.000000  4.223125e+04  
max      354.000000    137.000000   5657.000000  2.065902e+06

In [12]:
mean_hp_before = car_price_df['Engine HP'].mean()
print(round(mean_hp_before))
values = {'Engine HP': mean_hp_before}

249


In [13]:
updated_car_price_df = car_price_df.fillna(value=values)
updated_car_price_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 11914 entries, 0 to 11913
Data columns (total 16 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Make               11914 non-null  object 
 1   Model              11914 non-null  object 
 2   Year               11914 non-null  int64  
 3   Engine Fuel Type   11911 non-null  object 
 4   Engine HP          11914 non-null  float64
 5   Engine Cylinders   11884 non-null  float64
 6   Transmission Type  11914 non-null  object 
 7   Driven_Wheels      11914 non-null  object 
 8   Number of Doors    11908 non-null  float64
 9   Market Category    8172 non-null   object 
 10  Vehicle Size       11914 non-null  object 
 11  Vehicle Style      11914 non-null  object 
 12  highway MPG        11914 non-null  int64  
 13  city mpg           11914 non-null  int64  
 14  Popularity         11914 non-null  int64  
 15  MSRP               11914 non-null  int64  
dtypes: float64(3), int64(5

In [14]:
mean_hp_after = updated_car_price_df['Engine HP'].mean()
print(round(mean_hp_after))

249


Therefore, there is **NO change in the mean value**.

In [15]:
rolls_royce_df = car_price_df[car_price_df['Make'] == 'Rolls-Royce'][["Engine HP", "Engine Cylinders", "highway MPG"]]
unique_rows = rolls_royce_df.drop_duplicates()
print(unique_rows)

       Engine HP  Engine Cylinders  highway MPG
2921       325.0               8.0           15
3505       563.0              12.0           19
5275       563.0              12.0           21
5279       563.0              12.0           20
7443       322.0              12.0           15
7553       453.0              12.0           19
11448      624.0              12.0           21


In [16]:
X = np.array(unique_rows)
print(X)

[[325.   8.  15.]
 [563.  12.  19.]
 [563.  12.  21.]
 [563.  12.  20.]
 [322.  12.  15.]
 [453.  12.  19.]
 [624.  12.  21.]]


In [17]:
er = np.array([
    [1,2,3],
    [4,5,6],
    [5,6,7]
])
#np.linalg.inv(er)

In [18]:
XTX = X.T @ X
np.linalg.inv(XTX).sum()

0.032212320677486125

Therefore, the sum of all the elements of the result is : **0.0322**.

In [19]:
y = np.array([1000, 1100, 900, 1200, 1000, 850, 1300])

In [20]:
w = (np.linalg.inv(XTX) @ X.T) @ y
print(w)

[ 0.19989598 31.02612262 31.65378877]


Therefore, the first element of w is **0.1998**.